Xamarin.Forms.Platform.MacOS/Renderers/ScrollViewRenderer.cs

```diff
@@ -200,27 +191,77 @@ void UpdateBackgroundColor()
 
 		void UpdateContentSize()
 		{
-			if (_contentRenderer == null)
+			if (ContentView == null || ScrollView == null)
 				return;
-			var contentSize = ((ScrollView)Element).ContentSize.ToSizeF();
-			if (!contentSize.IsEmpty)
-				_contentRenderer.NativeView.Frame = new RectangleF(0, Element.Height - contentSize.Height, contentSize.Width,
-					contentSize.Height);
+
+            ContentView.Frame = new RectangleF(ContentView.Frame.X, ContentView.Frame.Y, Frame.Width, Frame.Height);
+            ResetNativeNonScroll();
 		}
 
+        private bool ResetNativeNonScroll( )
+        {
+            if (ScrollView == null || ContentView == null)
+                return false;
+
+            if (ScrollView.ScrollY <= 0.0f && ContentView.DocumentVisibleRect().Location.Y > 0.0f)
+            {
+                ContentView.ScrollToPoint(new CoreGraphics.CGPoint(0, 0));
+                return true;
+            }
+
+            return false;
+        }
+
 		[Export(nameof(UpdateScrollPosition))]
 		void UpdateScrollPosition()
 		{
-			var convertedPoint = (DocumentView as NSView)?.ConvertPointFromView(ContentView.Bounds.Location, ContentView);
-			if (convertedPoint.HasValue)
-				ScrollView.SetScrolledPosition(Math.Max(0, convertedPoint.Value.X), Math.Max(0, convertedPoint.Value.Y));
+            if (ScrollView == null)
+                return;
+            
+            if (ScrollView.ContentSize.Height >= ScrollView.Height)
+            {
+                RectangleF visibleRect = ContentView.DocumentVisibleRect();
+                if (visibleRect.Location.Y > -1)
+                    ScrollView.SetScrolledPosition(Math.Max(0, visibleRect.Location.X), Math.Max(0, ContentView.Frame.Height - visibleRect.Location.Y));
+            }
+            else
+                ResetNativeNonScroll();            
 		}
 
 		void ClearContentRenderer()
 		{
+            if ((ContentView as FlippedClipView) != null)
+                (ContentView as FlippedClipView).ContentRenderer = null;
+            
 			_contentRenderer?.NativeView?.RemoveFromSuperview();
 			_contentRenderer?.Dispose();
 			_contentRenderer = null;
 		}
+
+		private sealed class FlippedClipView : NSClipView
```

Comment:
Will do.  I was hesitant to move it into its own file because it kinda has a singular purpose and was minimal code.  The only time I could see reusing it would be if something changes at the ground level of the layout system and we flipped the entire macos nsviews to more closely resemble forms.  I should have changes up in the AM.